src/realm/db.cpp

```diff
@@ -78,304 +78,192 @@ namespace {
+    ReadCount& allocate_entry()
     {
-        return get(old_pos.load(std::memory_order_relaxed));
+        // grab entry from freelist:
+        REALM_ASSERT(freelist != nil);
+        auto& r = get(freelist);
+        freelist = r.older;
+
+        if (newest == nil) {
+            // list was empty
+            REALM_ASSERT(oldest == nil);
+            newest = oldest = index_of(r);
+            r.older = r.newer = nil;
+        }
+        else {
+            // add to existing list
+            auto& n = get(newest);
+            REALM_ASSERT(n.newer == nil);
+            r.newer = nil;
+            r.older = newest;
+            newest = index_of(r);
+            n.newer = newest;
+        }
+        return r;
     }
 
-    bool is_full() const noexcept
+    int index_of(const ReadCount& rc)
     {
-        uint_fast32_t idx = get(last()).next;
-        return idx == old_pos.load(std::memory_order_relaxed);
+        return static_cast<int>(&rc - data);
     }
 
-    uint_fast32_t next() const noexcept
+    void free_entry(int idx)
     {
-        // do not call this if the buffer is full!
-        uint_fast32_t idx = get(last()).next;
-        return idx;
-    }
-
-    ReadCount& get_next() noexcept
+        auto& r = get(idx);
+        if (idx == oldest)
+            oldest = r.newer;
+        if (idx == newest)
+            newest = r.older; // can this happen?
+        if (r.older != nil)
+            get(r.older).newer = r.newer;
+        if (r.newer != nil)
+            get(r.newer).older = r.older;
+
+        // add to freelist
+        r.newer = nil;
+        r.older = freelist;
+        freelist = idx;
+    }
+
+    // This method re-initialises the last used VersionList entry to hold a new entry.
+    // Precondition: This should *only* be done if the caller has established that she
+    // is the only thread/process that has access to the VersionList. It is currently
+    // called from init_versioning(), which is called by DB::open() under the
+    // condition that it is the session initiator and under guard by the control mutex,
+    // thus ensuring the precondition.
+    // It is most likely not suited for any other use.
+    ReadCount& init_versioning() noexcept
     {
-        REALM_ASSERT(!is_full());
-        return data[next()];
+        int num_entries = entries;
+        entries = 0;
+        newest = oldest = freelist = nil;
+        expand_to(num_entries);
+        ReadCount& r = allocate_entry();
+        r.count_live = r.count_frozen = 0;
+        return r;
     }
 
-    void use_next() noexcept
+    bool is_full() const noexcept
     {
-        atomic_dec(get_next().count); // .store_release(0);
-        put_pos.store(uint32_t(next()), std::memory_order_release);
+        return freelist == nil;
     }
 
-    void cleanup() noexcept
+    void purge_versions(uint64_t& oldest_v, uint64_t& oldest_live_v, TopRefMap& top_refs,
+                        int& num_unreachable) noexcept
```

Comment:
This function actually _can_ throw, specifically on the last line. I don't know how much we care about being resilient to small allocation failures, but I think I've seen us try to allow them before. I know mongo just made them process-fatal.